### PR TITLE
Set on-chain membership in a subnet when a reconfiguration is triggered

### PR DESCRIPTION
**Pull request by adlrocha**
> It can't be merged and properly tested until a bundle with https://github.com/consensus-shipyard/ipc-actors/pull/67 is included

Implements: https://github.com/consensus-shipyard/lotus/issues/125

### CODE REVIEW

File: chain/consensus/mir/state_manager.go [hunk 1]
```diff
@@ -314,6 +319,19 @@ func (sm *StateManager) ApplyTXs(txs []*requestpb.Request) error {
...
+	// message in block to update on-chain membership
+	if valSet != nil {
+		// FIXME: We should pick up the genesis address and not use the default one
+		// once we move into the user-defined gateway territory
+		reconfigMsg, err := membership.NewSetMembershipMsg(genesis.DefaultIPCGatewayAddr, valSet)
+		if err != nil {
+			return err
+		}
+		// Make it a message with an empty signature, these config messages are executed implicitly
+		msgs = append(msgs, reconfigMsg)
```
Comment (dnkolegov) [inline chain/consensus/mir/state_manager.go:332]: Is adding a message into msgs from the air a good idea? Would it not be better to send such a message as we do for config messages? 
Or we could add a message only if there are enough votes for it.

**Review by dnkolegov — COMMENTED** [review]
Added two questions

File: chain/consensus/mir/state_manager.go (hunk 1 above)
Comment (adlrocha) [reply chain/consensus/mir/state_manager.go:332]: Wait, I think I made a mistake, this should only be done when the vote a reconfiguration is completed, but this is not the case here, right? You may have config messages before there's consensus for the reconfiguration, right? We should move this somewhere else then :(

File: chain/consensus/mir/state_manager.go [hunk 2]
```diff
@@ -363,28 +380,31 @@ func (sm *StateManager) ApplyTXs(txs []*requestpb.Request) error {
...
-		return err
+		return nil, err
 	}
 
-	enoughVotes, err := sm.countVote(t.NodeID(msg.ClientId), &valSet)
+	enoughVotes, finished, err := sm.processVote(t.NodeID(msg.ClientId), &valSet)
 	if err != nil {
 		log.With("validator", sm.id).Errorf("failed to apply config message: %v", err)
-		return nil
+		// @denis: do we need to have this a `nil` error?
```
Comment (adlrocha) [inline chain/consensus/mir/state_manager.go:392]:
```suggestion
		// This error is not critical for the operation of the validator process, we should notify
		// the user but not kill the process. Returning an error here would exit the validator
		// process with failure. 
```

**Comment by adlrocha** [comment]
Just wanted to report that the commit to report the initial membership on-chain LGTM. Thanks! Let's catch the outstanding bug and get this merged :pray:

**Pull request merged**